fix: ensure UI consistency between deferred events

trueadm opened the pull request:
Fixes https://github.com/sveltejs/svelte/issues/11809 and fixes https://github.com/sveltejs/svelte/issues/11769.

When dealing with deferred events, we now ensure that:

- if we defer whilst deferred updates are already pending, and the yield is active (with a microtask in between) we first flush those to ensure the UI is up-to-date. Given we only defer updates around events, it means that another event has come before our deferred changes have been updated.
- if we defer whilst deferred updates are already pending, and the yield has started and the event that is active is trusted then we can also flush.
- if we have a deferred event and an action occurs, then we ensure that the next updates are on a microtask rather than a yield.

dummdidumm commented:
For the tests that fail, could a solution for these be: Schedule a microtask, and if the other yields happen within that microtask, it does not call flushSync and instead also lines them up?

```js
let scheduled = false;
function yield_effect() {
  if (!scheduled && effects_in_queue) {
    flushSync();
  } else {
    scheduled = true;
    queueMicrotask(() => scheduled = false);
    // add effects to queue in yield mode
  }
}
```

From my testing it seems that between `mouseup` and `click` there's a microtask happening so it should satisfy that case. Though it also seems that having too many listeners of the same type on an element would also take too long and a microtask happens between them, so not sure if it would work out.

I'm also wondering if the existance of `$effect.pre` effects should result in the yield mode being ignored and everything happening in a microtask as before instead, so that stuff like #11785 would work. We can think about this separately though.

trueadm commented:
@dummdidumm Yep that's exactly what I found too. We also need to account for sync events, which have no microtask between them. Such as when you do `button.click()`, it will sync fire the click and submit events together with no microtask.

dummdidumm commented on packages/svelte/src/internal/client/runtime.js:
```diff
@@ -722,11 +727,14 @@ function process_effects(effect, collected_effects) {
-	if (previous_scheduler_mode !== FLUSH_YIELD && is_yield_task_queued && !is_micro_task_queued) {
+	// If we're calling yield_updates and there is already an active yield in progress (is_yield_task_active)
+	// then it's likely that the event might might try and read from the UI. Additionally, we might be dealing with
+	// event that was sync called from another event flow, thus isTrusted() will be true – such as a submit event from
+	// a subsequent click event. In either case, the UI needs to be up-to-date, so we flush any pending changes.
```
Can you elaborate a bit why isTrusted is a good hint for "event was sync called from another event flow"? From my understanding isTrusted only means "the user did this". How can we infer the sync-called state from it?

trueadm replied:
Well we can't fully, we can only infer that given that `is_yield_task_queued` is true, that some event before it has run and now if we have another event that is trusted it might have been called from that. Whilst the `click` won't be trusted, as it was done via `element.click()` the `submit` will be trusted.

trueadm commented:
I refactored the code some more, found another test case that needed fixing and also added more test coverage. I'm feeling quite good about the inner workings of this now, even if it gets a bit complicated. I still firmly believe this is something Svelte should handle. The amount of depth required from me to understand this is simply not something we should be pushing on our users.

trueadm commented on packages/svelte/src/internal/client/dom/elements/events.js:
```diff
@@ -173,23 +185,35 @@ export function handle_event_propagation(handler_element, event) {
+			var delegated = get_delegated(next_target, event_name);
+
+			if (delegated !== undefined) {
+				yield_event_updates(() => {
```
Because that's how events work natively – and this was a bug before. We should yield during propagation – I have a test that demonstrates this.

trueadm replied:
I changed this back to what it was before.

trueadm commented on packages/svelte/tests/runtime-runes/samples/deferred-events-consistency-3/_config.js:
```diff
@@ -0,0 +1,12 @@
+		b1.click();
+		flushSync();
+
+		assert.deepEqual(logs, ['works']);
```
We would have to undo the yield on each update, and do it on all events like before if we agreed we didn't want this change.

trueadm replied:
Also, if you don't spread the properties, the update still works properly in Svelte 5. It's only a bug when you spread.

trueadm replied:
better-deferred-heuristic

dummdidumm replied:
The bug comes from the spread attributes getting updated in a render effect but the static one changing based on a derived which is read at the time of invocation, which means it's up to date at that point. Clearly an inconsistency, and not sure how to fix that without using your approach after all, mhm ... @Rich-Harris thoughts on this?

trueadm commented on packages/svelte/src/internal/client/runtime.js:
```diff
@@ -709,10 +723,21 @@ function process_effects(effect, collected_effects) {
+ * @param {boolean} [is_trusted]
  */
-export function yield_updates(fn) {
+export function yield_event_updates(fn, is_trusted = false) {
```
That sounds like a good idea, feel free to play around with it on this branch. I'm looking at something else today.

trueadm replied:
Okay I changed the test and added a TODO. We can fix this in another PR.

trueadm replied:
`component-transition` test
